include/picongpu/plugins/xrayScattering/XrayScattering.hpp

```diff
@@ -0,0 +1,846 @@
+                T_ParticlesType >::type;
+            // Calculate density.
+            tmpField-> template computeValue<
+                CORE + BORDER,
+                ElectronDensitySolver
+            >(
+                * species,
+                currentStep
+            );
+            // Release particle data.
+            dc.releaseData( T_ParticlesType::FrameType::getName( ) );
+            // Get the field data box.
+            FieldTmp::DataBoxType tmpFieldBox =
+                tmpField->getGridBuffer( ).getDeviceBuffer( ).getDataBox( );
+            return tmpFieldBox;
+        }
+
+
+        /** Runs xrayScattering kernel when the output is split over nodes.
+         *
+         * A single kernel run adds result only to that output part which
+         * currently resides on the node. The output parts are passed along to
+         * the neigbouring node, in a circle. This repeats until every node has
+         * computed all scattering vectors.
+         *
+         * @param cellsGrid field grid, without GUARD, on one device
+         * @param fieldTmpNoGuard field data
+         * @param globalOffset offset from the global to the local domain
+         * @param numBlocks number of virtual blocks used in a kernel run
+         * @param fieldPos TmpField in cell position
+         */
+        template < typename T_FieldPos >
+        HINLINE void runKernelInSplitMode(
+            DataSpace < simDim > & cellsGrid,
+            FieldTmp::DataBoxType const & fieldTmpNoGuard,
+            DataSpace< simDim > & globalOffset,
+            uint32_t const & numBlocks,
+            T_FieldPos const & fieldPos
+        )
+        {
+            // The available number of virtual workers.
+            constexpr uint32_t numWorkers = pmacc::traits::GetNumWorkers<
+                pmacc::math::CT::volume< SuperCellSize >::type::value >::value;
+
+            uint32_t countVectors, iterOffset;
+            // Loop over kernel runs.
+            for( uint32_t step = 0; step < countRanks; step++ )
+            {
+                // Pass along the data.
+                communicationOnStep( step );
+                // 1D offset to the begin of the currently processed output
+                // part.
+                iterOffset = calcOffset( step );
+                // Define scattering vectors for the output part.
+                GetScatteringVector scatteringVectors
+                    {
+                        q_min,
+                        q_max,
+                        q_step,
+                        numVectors,
+                        iterOffset
+                    };
+                // Handle possibly smaller amount of vectors to be processed
+                // in the last output part.
+                if ( hasLastChunk( step ) )
+                {
+                    countVectors =  resOfVectors;
+                }
+                else countVectors = amplitude->getHostBuffer(
+                    ).getCurrentSize( );
```

Comment:
Perhaps it would be a little cleaner to define variable `countVectors` inside the loop and use the `?` operator to assign the initial value to it instead of `if`.

Reply:
I moved the declaration into the loop.

 I think the rule in our style guidelines was to use `? :` when it can fit a single line. I'm trying to keep my code lines <= 80 characters. But yes, this makes it a bit lengthy. To be honest, I don't know what would be a better practise. @sbastrakov

Reply:
It's fine, nothing wrong with keeping the `if` statement.